### ISSUE TITLE
MIrror ftp5.gwdg.de - prioritize planet pbf

### ISSUE THREAD

**Issue by mmd-osm** [openstreetmap/operations#364]
As we've seen in the last couple of weeks, mirror ftp5.gwdg.de might take up to 1.5 days to replicate new pbf/osm.bz2 planet and history files. Unfortunately, we have to wait for those >300GB to be fully replicated before we can redirect a single file. Ideally, the mirror should replicate the planet pbf files first and make them available as early as possible.

My suggestion is to contact https://www.gwdg.de/support via their email support address, and discuss with them if they're able to adjust their replication schema to better suit our needs.

**Comment by iandees** [comment]
I'm working with the OSUOSL mirroring folks (http://ftp.osuosl.org/pub/openstreetmap/) as well. They're asking for a rsync module/incantation that they should use to perform the mirroring. Maybe this is something we can share so more mirroring teams can do the same thing?

**Comment by tomhughes** [comment]
Sure - they can use `planet.openstreetmap.org::planet` which is mapped to the root of tree that you can see at https://planet.openstreetmap.org/.

**Comment by Firefishy** [comment]
If rsyncing, likely a good idea to `--exclude *-latest.*` too.

**Comment by RubenKelevra** [comment]
Sorry for hijacking this ticket - wasn't comfortable to create a new ticket 'demanding' any change on established practices... but!

I was just wondering if it's not worth considering a system change, rsync isn't a great solution in a star-topology. Especially if you want to cut down the delay and don't want to present outdated files to the users (which picked the wrong mirror server).

How about changing this infrastructure towards an IPFS-Cluster?

Using IPFS does remove all issues with the current infrastructure:

It can reliably hold a globally mirrored dataset with instant access for the users on the completion of new files. The switchover from one 'version' to the next of a directory is atomic and it's guaranteed that the content of a directory doesn't change while you're downloading/accessing/using it.

To set up a cluster you just have to provide one server that will be allowed to write on the cluster. If you have multiple machines that need to write to the cluster - that's also possible.

Everyone whos willing to hold data of the cluster can 'follow' it. You can limit the number of copies of each element hold in the cluster, say you just want to start with 15 copies of each file and if it drops below 10 copies an additional copies should be made.

There's a desktop client for Linux/Windows/MacOS, there are web-gateways provided by the project itself and [cloudflare](https://blog.cloudflare.com/distributed-web-gateway/).

Additionally [Netflix did major improvements](https://blog.ipfs.io/2020-02-14-improved-bitswap-for-container-distribution/) on the data exchange protocol (bitswap) which is going to be released on the next major version. 

For archiving purposes, you can rearrange the files to your likings on any IPFS node, since the content is addressed and automatically deduped you just create some additional metadata around the actual files.

In the end, I like to thank you for reading to the end and all the years of great service! :)

[an example page stored on IPFS](https://ipfs.io/ipfs/QmVMntAvXr9AjgGhdGNDbnZR2vZKk38QNh6hZqDLuWgsdp/)

**Comment by iandees** [comment]
I forgot about this ticket and Lance from OSUOSL asked for an update last week.

I'm trying to suggest a reasonable rsync stanza to use. For some reason I thought there was a time limit/filter on rsync so that I could suggest "grab everything from the last 10 days", but that doesn't seem to be there.

If I'm trying to suggest grabbing the most recent planet files in the two formats, what's a good rsync command to use?

**Comment by mmd-osm** [comment]
@RubenKelevra : I don't see how this is relevant for the topic at hand. If you want to propose some entirely new approach, please create a new issue rather than hijacking some unrelated issue.

**Comment by RubenKelevra** [comment]
> I forgot about this ticket and Lance from OSUOSL asked for an update last week.
> 
> I'm trying to suggest a reasonable rsync stanza to use. For some reason I thought there was a time limit/filter on rsync so that I could suggest "grab everything from the last 10 days", but that doesn't seem to be there.
> 
> If I'm trying to suggest grabbing the most recent planet files in the two formats, what's a good rsync command to use?

Well, you could just exclude all files you don't want to sync. A timestamp based selection on the other hand is AFAIK just possible if you have ssh access to run a `find` on the remote server first, to build the filelist you like to sync.

As I said, Rsync isn't ideal in this kind of setup, because it can't speedup syncs by interconnecting the destination servers to each other and exchange data.

I gonna create a new ticket for a discussion how to change away from the current rsync approach to a ipfs cluster.

**Comment by mmd-osm** [comment]
OSM mirrors are provided by a number of different third parties that are pretty much independent of each other. I think it's highly unlikely that they want to set up some kind of interconnect as you describe.

If you're looking into some kind of p2p protocol, you should check https://twitter.com/OSM_Tech/status/1222507257478422535 instead which is already available today.

**Comment by RubenKelevra** [comment]
Well, with an IFPS cluster, you can let those third parties join it and hold a copy on their mirror, receiving all changes you do the cluster.

Instead of running a rsync from now and then, they would run an IPFS instance with a cluster follower. Since the content is addressed, and not the location an outage of a third party doesn't matter. If it gets back online, the cluster follower would receive the delta of the changes and fetch all new content (and remove the old one).

Compared to pure IPFS the cluster will also guarantee that each file is at least x times available on the cluster, so each follower doesn't have to hold the full dataset - if there are enough followers available the data will be shared.

As I said, I create a new ticket about that, pointing out the pros and cons and limitations of IPFS.

**Comment by mnalis** [comment]
As we have [BitTorrent](https://github.com/openstreetmap/operations/issues/451) implemented now, it would probably be fastest and most efficient (and put the least load on OSM) for mirrors to rsync only `.torrent` files to directory which automatically starts downloading them (for example, by using `transmission-daemon`, `rtorrent` or `aria2` or some other [torrent client](https://wiki.openstreetmap.org/wiki/Planet.osm#BitTorrent)).
